### PR TITLE
Add support of the command COPY

### LINKED ISSUE #1945: Add support of the command COPY
Refer to https://redis.io/commands/copy/

### PR DESCRIPTION
**Pull request by Chiro11**
Close #1945 

DB number can only be 0 in this command, which doesn't mean anything.
Copying between namespaces will be implemented in another command called `COPYX` in later pr.

Note: most tests are modified from `rename_test.go`

### CODE REVIEW

File: src/storage/redis_db.cc [hunk 1]
```diff
@@ -708,7 +708,7 @@ rocksdb::Status Database::typeInternal(const Slice &key, RedisType *type) {
   return rocksdb::Status::OK();
 }
 
-rocksdb::Status Database::Move(const std::string &key, const std::string &new_key, bool nx, bool *ret,
+rocksdb::Status Database::Move(const std::string &key, const std::string &new_key, bool nx, bool delete_old, bool *ret,
                                bool *key_exist) {
```
Comment (PragmaTwice) [inline src/storage/redis_db.cc]: Please unify `ret` and `key_exist` to just one output value.

File: src/storage/redis_db.cc [hunk 2]
```diff
@@ -708,7 +708,7 @@ rocksdb::Status Database::typeInternal(const Slice &key, RedisType *type) {
   return rocksdb::Status::OK();
 }
 
-rocksdb::Status Database::Move(const std::string &key, const std::string &new_key, bool nx, bool *ret,
+rocksdb::Status Database::Move(const std::string &key, const std::string &new_key, bool nx, bool delete_old, bool *ret,
```
Comment (PragmaTwice) [inline src/storage/redis_db.cc]: We can rename it to `Copy`.

File: src/storage/redis_db.cc (hunk 1 above)
Comment (Chiro11) [reply src/storage/redis_db.cc]: How to do it, use int?

Comment (PragmaTwice) [reply src/storage/redis_db.cc]:
```
enum { KEY_NOT_EXIST, KEY_ALREADY_EXIST, DONE }
```

**Review by jihuayu — COMMENTED** [review]
I found sonar cloud tells us "Rename function "TestCopy_Error" to match the regular expression ^(_|[a-zA-Z0-9]+)$".
Can you change those go test case function names to pascal Case (eg: `TestCopyError`)?

https://sonarcloud.io/project/issues?resolved=false&pullRequest=2238&id=apache_kvrocks&open=AY7N7bb-6GZZ34cBM2MJ

**Comment by Chiro11** [comment]
> I found sonar cloud tells us "Rename function "TestCopy_Error" to match the regular expression ^(_|[a-zA-Z0-9]+)$". Can you change those go test case function names to pascal Case (eg: `TestCopyError`)?
> 
> https://sonarcloud.io/project/issues?resolved=false&pullRequest=2238&id=apache_kvrocks&open=AY7N7bb-6GZZ34cBM2MJ

OK. I will also change the function name in rename_test.go by the way.

File: src/commands/cmd_key.cc [hunk 3]
```diff
@@ -222,7 +221,7 @@ class CommandExpireAt : public Commander {
 
     timestamp_ = *parse_result;
 
-    return Commander::Parse(args);
+    return Status::OK();
```
Comment (git-hulk) [inline src/commands/cmd_key.cc:224]: Why do we change this line?

File: src/commands/cmd_key.cc [hunk 4]
```diff
@@ -363,20 +361,80 @@ class CommandRenameNX : public Commander {
...
+        replace_ = true;
+      } else {
+        return {Status::RedisParseErr, errInvalidSyntax};
+      }
+    } else if (args.size() >= 5) {
+      if (util::ToLower(args[3]) != "db") {
+        return {Status::RedisParseErr, errInvalidSyntax};
+      }
+      auto db_num = ParseInt<int64_t>(args[4], 10);
+      // DB num should only be 1, just a placeholder
```
Comment (git-hulk) [inline src/commands/cmd_key.cc]:
```suggestion
      // There's only one database in Kvrocks, so the DB must be 0 here.
```

File: src/commands/cmd_key.cc (hunk 3 above)
Comment (Chiro11) [reply src/commands/cmd_key.cc:224]: Just to unify. Because the other functions in this file are written like this.

Comment (git-hulk) [reply src/commands/cmd_key.cc:224]: ok, that makes sense since it does nothing inside the parent's Parse().

**Comment by Chiro11** [comment]
@git-hulk One ci test failed. Did it happen occasionally or something wrong?

**Comment by git-hulk** [comment]
@Chiro11 Yes, it's unrelated to this PR and there's no worry about it. It seems a bug inside the llvm compiler: https://github.com/apache/kvrocks/actions/runs/8657998616/job/23741619462?pr=2238#step:11:1632

File: src/commands/cmd_key.cc [hunk 5]
```diff
@@ -363,20 +361,80 @@ class CommandRenameNX : public Commander {
...
+  }
+};
+
+class CommandCopy : public Commander {
+ public:
+  Status Parse(const std::vector<std::string> &args) override {
+    if (args.size() > 6) {
+      return {Status::RedisParseErr, errWrongNumOfArguments};
+    }
+    if (args.size() == 4) {
```
Comment (git-hulk) [inline src/commands/cmd_key.cc]: From the Redis documentation: https://redis.io/docs/latest/commands/copy/, it doesn't require the DB to be placed before REPLACE. That said: `COPY k0 k1 REPLACE DB 0` is also good for Redis, you can implement the parameter parser via `CommandParser`.

**Comment by git-hulk** [comment]
@Chiro11 This PR generally looks good for me and one comment is inline. Thanks for your careful code coverage for this command, that's cool.

File: src/commands/cmd_key.cc [hunk 6]
```diff
@@ -346,14 +345,13 @@ class CommandRename : public Commander {
...
+    Database::CopyResult res = Database::CopyResult::DONE;
     std::string ns_key = redis.AppendNamespacePrefix(args_[1]);
     std::string new_ns_key = redis.AppendNamespacePrefix(args_[2]);
-    auto s = redis.Move(ns_key, new_ns_key, false, &ret, &key_exist);
+    auto s = redis.Copy(ns_key, new_ns_key, false, true, &res);
     if (!s.ok()) return {Status::RedisExecErr, s.ToString()};
-    if (!key_exist) return {Status::RedisExecErr, rocksdb::Status::InvalidArgument("ERR no such key").ToString()};
-
+    if (res == Database::CopyResult::KEY_NOT_EXIST)
+      return {Status::RedisExecErr, rocksdb::Status::InvalidArgument("ERR no such key").ToString()};
```
Comment (PragmaTwice) [inline src/commands/cmd_key.cc]:
```suggestion
      return {Status::RedisExecErr, "no such key"};
```
rocksdb::Status is meaningless here..

File: src/commands/cmd_key.cc [hunk 7]
```diff
@@ -363,20 +361,80 @@ class CommandRenameNX : public Commander {
...
+      if (!db_num.IsOK() || db_num.GetValue() != 0) {
+        return {Status::RedisParseErr, errInvalidSyntax};
+      }
+      if (args.size() == 6) {
+        if (util::ToLower(args[5]) == "replace") {
+          replace_ = true;
+        } else {
+          return {Status::RedisParseErr, errInvalidSyntax};
+        }
+      }
```
Comment (PragmaTwice) [inline src/commands/cmd_key.cc]: Please rewrite via `CommandParser`.

File: src/commands/cmd_key.cc (hunk 5 above)
Comment (PragmaTwice) [reply src/commands/cmd_key.cc]: Also, we don't need to have a `COPYX`, we can just extend this command by `COPY ... NAMESPACE xx`.

**Review by PragmaTwice — COMMENTED** [review]
Looks good to me now.

> Also, we don't need to have a COPYX, we can just extend this command by COPY ... NAMESPACE xx.

BTW do you want to implement it in this PR?

**Comment by Chiro11** [comment]
> Looks good to me now.
> 
> > Also, we don't need to have a COPYX, we can just extend this command by COPY ... NAMESPACE xx.
> 
> BTW do you want to implement it in this PR?

I prefer put it in another PR. It also need tests, make this PR too huge.

**Review by PragmaTwice — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Pull request merged**